sandbox/lua/decoders/syslog.lua

```diff
@@ -0,0 +1,154 @@
+
+    [SyslogDecoder.config]
+    type = 'RSYSLOG_TraditionalFileFormat'
+    rsyslog_template = '%TIMESTAMP% %HOSTNAME% %syslogtag%%msg:::sp-if-no-1st-sp%%msg:::drop-last-lf%\n'
+    tz = 'Europe/Paris'
+
+*Example Heka Message*
+
+:Timestamp: 2014-01-10 07:04:56 -0800 PST
+:Type: RSYSLOG_TraditionalFileFormat
+:Hostname: test.example.org
+:Pid: 0
+:UUID: 8e414f01-9d7f-4a48-a5e1-ae92e5954df5
+:Logger: SyslogInput
+:Payload: 25F2E5E061: to=<george.desantis@test.example.com>, relay=none, delay=0.05, delays=0.05/0/0/0, dsn=2.0.0, status=sent (test.example.com)
+:EnvVersion:
+:Severity: 7
+:Fields:
+    | name:"programname" value:"postfix/discard"
+    | name:"postfix_queueid" value:"25F2E5E061"
+    | name:"postfix_to" value:"george.desantis@test.example.com"
+    | name:"postfix_relay" value:"none"
+    | name:"postfix_delay" value:0.05
+    | name:"postfix_delay_before_qmgr" value:0.05
+    | name:"postfix_delay_in_qmgr" value:0
+    | name:"postfix_delay_conn_setup" value:0
+    | name:"postfix_delay_transmission" value:0
+    | name:"postfix_dsn" value:"2.0.0"
+    | name:"postfix_status" value:"sent"
+--]]
+
+local l = require 'lpeg'
+local syslog = require 'syslog'
+local syslog_message = require 'syslog_message'
+local postfix = require 'postfix'
+
+local rsyslog_template = read_config('rsyslog_template')
+local msg_type = read_config('type')
+local hostname_keep = read_config('hostname_keep')
+
+local msg = {
+    Timestamp   = nil,
+    Type        = msg_type,
+    Hostname    = nil,
+    Payload     = nil,
+    Pid         = nil,
+    Severity    = nil,
+    Fields      = nil
+}
+
+local rsyslog_grammar = syslog.build_rsyslog_grammar(rsyslog_template) or error('Unable to parse rsyslog template')
+
+local wildcard_grammars = syslog_message.get_wildcard_grammar()
+
+function process_message ()
+    local log = read_message('Payload')
+    local fields = rsyslog_grammar:match(log)
+    if not fields then return -1 end
+    if fields.timestamp then
+        msg.Timestamp = fields.timestamp
+        fields.timestamp = nil
+    end
+
+    if fields.pri then
+        msg.Severity = fields.pri.severity
+        fields.syslogfacility = fields.pri.facility
+        fields.pri = nil
+    else
+        msg.Severity = fields.syslogseverity or fields['syslogseverity-text']
+        or fields.syslogpriority or fields['syslogpriority-text']
+
+        fields.syslogseverity = nil
+        fields['syslogseverity-text'] = nil
+        fields.syslogpriority = nil
+        fields['syslogpriority-text'] = nil
+    end
+
+    if fields.syslogtag then
+        fields.programname = fields.syslogtag.programname
+        msg.Pid = fields.syslogtag.pid
+        fields.syslogtag = nil
+    end
+
+    if not hostname_keep then
+        msg.Hostname = fields.hostname or fields.source
+        fields.hostname = nil
+        fields.source = nil
+    end
+
+    msg.Payload = fields.msg
+    fields.msg = nil
+
+    -- Now parses syslog msg
```

Comment:
It could, but the memory impact is not minor.

But the patch can be reworked to only require 'syslog_message' and 'postfix' when needed.